### PR TITLE
Remove class VectorSpaceVector.

### LINKED ISSUE #3073: Remove VectorSpaceVector.
What are the plans for how to solve operations on a VectorSpaceVector that internally needs to access individual elements? Although we want to hide explicit access of individual elements from the outside, we certainly need to do so inside some operations, such as a sparse-matrix vector product. As I see it, the natural way to do this would be to let SparseMatrix be a friend to the vector classes. Are there any other superior solutions?

I am working on a sparse matrix implementation for GPU, and there the `vmult` operation needs to extract the underlying data pointer and pass it to a CUDA kernel. Also this would be solved by a friend relation.

### PR DESCRIPTION
**Pull request by bangerth**
Final step for #3073. Closes #3073.

### CODE REVIEW

**Review by kronbichler — APPROVED** [review]

**Review by marcfehling — APPROVED** [review]

**Comment by marcfehling** [comment]
```
/home/runner/work/dealii/dealii/source/lac/cuda_vector.cc:129:21: error: out-of-line definition of 'import_elements' does not match any declaration in 'Vector<Number>'
    Vector<Number>::import_elements(
                    ^~~~~~~~~~~~~~~
```

**Comment by bangerth** [comment]
I cannot figure out why the CUDA compiler is unhappy. The first commit does not actually touch the CUDA vector class at all, but we get the same error as pointed out above. That should not have happened in the first place.

So I added the last commit here that applies a similar patch as I've made in other places. The CUDA compiler complains that this definition
https://github.com/bangerth/dealii/blob/vsv-final/source/lac/cuda_vector.cc#L127-L132
does not match any declaration, and in particular not this one:
https://github.com/bangerth/dealii/blob/vsv-final/include/deal.II/lac/cuda_vector.h#L137-L142
Does anyone see what I'm apparently missing?

File: include/deal.II/lac/vector_space_vector.h [hunk 1]
```diff
@@ -1,304 +0,0 @@
...
-   * Shift all entries of the vector by a constant factor so that the mean
-   * value of the vector becomes zero.
-   */
-  template <typename Number>
-  void
-  set_zero_mean_value(VectorSpaceVector<Number> &vector)
-  {
-    vector.add(-vector.mean_value());
-  }
-} // namespace LinearAlgebra
```
Comment (kronbichler) [inline include/deal.II/lac/vector_space_vector.h:300]: This is something we should move somewhere else. This change will break compilation of https://github.com/exadg/exadg for example. We should move this function somewhere else with `template <typename VectorType>` arguments, the question is where we want it to be to minimize backwards compatibility break.

Comment (bangerth) [reply include/deal.II/lac/vector_space_vector.h:300]: Good point. I could just put that back into this file as the only thing that's in there. Would that be acceptable?

**Comment by Rombur** [comment]
@bangerth the forward declaration of `ReadWriteVector` in `cuda_vector.h` is wrong. It needs to be put inside the `LinearAlgebra` namespace. That fixed the issue for me.

**Comment by bangerth** [comment]
Oh, good detective work! What a bizarre failure mode!

File: include/deal.II/lac/vector_space_vector.h (hunk 1 above)
Comment (kronbichler) [reply include/deal.II/lac/vector_space_vector.h:300]: Do we have another file which is included by all vector classes? I think many use cases did just include the derived vector class, e.g. `deal.II/lac/la_parallel_vector.h` in our case of ExaDG, and got the declaration in `vector_space_vector.h` on the way. I'm not sure where it fits best, but the three options are `read_vector.h`, `vector_type_traits.h` and `vector_operation.h`, none of which uses the `LinearAlgebra` namespace. We could also set a fourth file `vector_templace_functions.h` or similar that all our vector classes include.

**Comment by Rombur** [comment]
> Oh, good detective work! What a bizarre failure mode!

That's nvcc for you. Once I forgot to implement a function, the code compiled but segfaulted at runtime...

**Review by masterleinad — APPROVED** [review]

File: include/deal.II/lac/vector_space_vector.h (hunk 1 above)
Comment (drwells) [reply include/deal.II/lac/vector_space_vector.h:300]: `VectorTools::subtract_mean_value()` does the same thing:

https://github.com/dealii/dealii/blob/de6a2e7a6f4ffdc91b780e67b5e468f921f5a27c/include/deal.II/numerics/vector_tools_mean_value.templates.h#L90-L98

We could add the header back in but deprecate the function in favor of the `VectorTools` version.

Comment (kronbichler) [reply include/deal.II/lac/vector_space_vector.h:300]: I agree.

**Comment by bangerth** [comment]
Oh, we merged this without addressing the function that needs to be put in. The easy way is to just keep around the file for that one functino. @kronbichler Would that be ok with you?

**Comment by kronbichler** [comment]
Yes, keeping the file would be ok, thank you.

**Comment by bangerth** [comment]
See #15715.

**Pull request merged**